### PR TITLE
Add UpdateCommand

### LINKED ISSUE #96: <cli> sidekick upgrade
Create an `upgrade` command that updates the bash scripting of the already installed cli

### PR DESCRIPTION
**Pull request by Giuspepe**
Fixes #96 

## Scope of this PR
- `pubspec.yaml` template now contains a field `sidekick: cli_version: <version>` which contains the `sidekick` version the CLI was generated with
- Checks whether an update is available whenever a command is executed
- Add `UpdateCommand` which calls the latest available version of `update_sidekick_cli.dart`
- `update_sidekick_cli` currently updates the following:
  - shell scripts in `tool` (`download_dart.sh`, `install.sh`, `run.sh`,  `sidekick_config.sh`)
  - entrypoint executable
  - `sidekick_core` dependency to latest available version
  - `sidekick->cli_version`  in `pubspec.yaml` to latest available `sidekick` version

### Update `sidekick_core` in `sidekick` and pass new sidekick version parameter to `SidekickTemplateProperties`
When generating a new sidekick CLI package, we have to save the version of sidekick which generated the CLI in the generated CLI's `<generated-cli>/pubspec.yaml`. However, we can't read the global sidekick's version because its executable is compiled and thus does not have access to its own `<global-sidekick>/pubspec.yaml` because it isn't bundled with the compiled executable.  
Instead we could use an automatically generated source file which contains the version information
See e.g. https://stackoverflow.com/a/69048343/9905602 or how mason_cli implements this in a version.dart file: https://github.com/felangel/mason/pull/538/files

We should save the version of `sidekick_core` from `sidekick`'s `pubspec.lock` in a Dart file

### Add migration functions
Whenever additional steps (besides regenerating shell scripts) are needed to migrate from an old to the latest version, add them to `update_sidekick_cli.dart`. This script receives the current and the latest version. This information can be used to call migration functions to do additional update steps (e.g. modifying certain lines in dart files).

## How to test this PR
There is a `update_command_test.dart` which already tests this command.
If you want to test this manually, do the following:
1. Edit the `pubspec.yaml` of an old sidekick CLI

```yaml
dependencies:
  sidekick_core: '>=0.11.0 <1.0.0'

dependency_overrides:
  sidekick_core:
    path: /Users/pepe/repos/sidekick/sidekick_core
```
2. Add `SidekickCommand` to the cli
3. Run `<cli> sidekick update`
4. Verify the following:  
    1. the 4 shell scripts in `tool` have been updated  
    2. `cli_version` has been updated to the latest version of `sidekick` (`0.7.1`)  
    3.  the `sidekick_core` dependency has been updated to `sidekick_core: '>=0.11.1 <1.0.0'`
    4. A success message has been printed (`Successfully updated sidekick CLI <cli> from version 0.0.0 to 0.7.1!`)

### CODE REVIEW

**Comment by Giuspepe** [comment]
Maybe we should also uses this command to update the bundled dart runtime (`sidekickDartRuntime`). See https://github.com/phntmxyz/sidekick/issues/57

**Comment by Giuspepe** [comment]
When running `<cli> sidekick update` as described above in an outdated sidekick CLI almost everything works. This is the output:
```
Updating sidekick CLI iv from version 0.0.0 to 0.7.1 ...
Successfully updated sidekick CLI iv from version 0.0.0 to 0.7.1!
/Users/pepe/repos/invo-frontend/packages/iv_sidekick/tool/run.sh: line 70: unexpected EOF while looking for matching `"'
``` 
Everything is correctly updated and executed from the dart side, but an error (exit code 2) is thrown from the `run.sh` script: 

```
/Users/pepe/repos/invo-frontend/packages/iv_sidekick/tool/run.sh: line 70: unexpected EOF while looking for matching `"'
```

I guess because we're modifying the file which is currently being executed?  🤔

File: sidekick_core/lib/src/commands/update_command.dart [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  final String description = 'Updates the sidekick cli';
+
+  @override
+  final String name = 'update';
+
+  @override
+  Future<void> run() async {
+    const sidekickVersionChecker = SidekickVersionChecker();
+    final latestSidekickVersion =
+        await sidekickVersionChecker.getLatestPackageVersion('sidekick');
```
Comment (passsy) [inline sidekick_core/lib/src/commands/update_command.dart:16]: We might ship updates to package `sidekick` that do not change the generated code. Maybe we should create a `template_version`

File: sidekick_core/pubspec.yaml [hunk 2]
```diff
@@ -11,7 +11,12 @@ dependencies:
   args: ^2.1.0
   dartx: '>=0.7.1 <2.0.0'
   dcli: ^1.15.0
+  deep_pick: ^0.10.0
+  http: ^0.13.5
   meta: ^1.5.0
+  pub_semver: ^2.1.3
+  # transitive from dcli
+  pubspec2: ^2.4.1
```
Comment (passsy) [inline sidekick_core/pubspec.yaml:19]: We should be very cautious when adding new packages to the core because it is shipped with every sidekick cli.

For sure we need `http`.
`pub_semver` + `pubspec2` are already in `dcli`, fine.

But we can easily get rid of `deep_pick`.

File: sidekick_core/test/update_command_test.dart [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+name: dash
+
+environment:
+  sdk: '>=2.14.0 <3.0.0'
+
+dependencies:
+  sidekick_core: 0.0.0
+
+sidekick:
+  generator_version: 0.0.0
```
Comment (passsy) [inline sidekick_core/test/update_command_test.dart:105]: We don't need `generator_version` to be a semver, a simple int is sufficient

Comment (Giuspepe) [reply sidekick_core/test/update_command_test.dart:105]: I read the latest version from pub.dev which is a semver and very easy compared to reading the latest version from the right file in the github repository at the right revision

File: sidekick_core/lib/src/sidekick_version_checker.dart [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    final newVersionConstraint = pinVersion
+        ? newMinimumVersion.canonicalizedVersion
+        : newMinimumVersion.major > 0
+            ? '^${newMinimumVersion.canonicalizedVersion}'
+            : "'>=${newMinimumVersion.canonicalizedVersion} <1.0.0'";
+
+    final index = lines.indexWhere((it) => it.startsWith('  $package:'));
+    lines[index] = '  $package: $newVersionConstraint';
+
+    pubspec.writeAsStringSync(lines.join('\n'));
```
[the remark this answers is not in the document]
Comment (Giuspepe) [reply sidekick_core/lib/src/sidekick_version_checker.dart:58]: Ah that gave me the idea to use `dart pub remove <package> && dart pub add <package>:<version-constraint>` instead :)

[the remark this answers is not in the document]
Comment (Giuspepe) [reply sidekick_core/lib/src/sidekick_version_checker.dart:58]: Nevermind I'll leave it as is because otherwise I can't reuse this method to update `sidekick: cli_version` in the pubspec.yaml since that isn't a dependency

File: sidekick_core/lib/src/sidekick_version_checker.dart [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      '[\'"\\^<>= ]*(\\d+\\.\\d+\\.\\d+(?:[+-]\\S+)?)',
+    );
+
+    final minVersion = versionConstraintRegEx
+        .allMatches(versionConstraint)
+        .map((e) => e.group(1))
+        .whereNotNull()
+        .first;
+
+    return Version.parse(minVersion);
```
[the remark this answers is not in the document]
Comment (Giuspepe) [reply sidekick_core/lib/src/sidekick_version_checker.dart:95]: I tried doing that but didn't like it because you end up needing something like this:

```dart
final pubspec =PubSpec.fromFile(Repository.requiredSidekickPackage.pubspec.path);
return (pubspec.dependencies['sidekick_core']?.reference as ExternalHostedReference).versionConstraint as VersionRange).min;
```

With that approach I also ran into this error straight from dependency hell:
```
The name 'PubSpec' is defined in the libraries 'package:dcli/src/pubspec/pubspec.dart' and 'package:pubspec2/src/pubspec.dart (via package:pubspec2/pubspec2.dart)'
```

File: sidekick_core/lib/src/sidekick_version_checker.dart [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    final pubspec =
+        loadYaml(Repository.requiredSidekickPackage.pubspec.readAsStringSync());
+
+    // ignore: avoid_dynamic_calls
+    dynamic current = pubspec[path.first];
+    for (final key in path.sublist(1)) {
+      // ignore: avoid_dynamic_calls
+      current = current[key];
```
Comment (passsy) [inline sidekick_core/lib/src/sidekick_version_checker.dart:110]: Throws when the `pubspec.yaml` doesn't contain the `sidekick` property

```
NoSuchMethodError: The method '[]' was called on null.
Receiver: null
Tried calling: []("cli_version")
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:38)
#1      _objectNoSuchMethod (dart:core-patch/object_patch.dart:85)
#2      SidekickVersionChecker._readFromPubspecYaml (package:sidekick_core/src/sidekick_version_checker.dart:110)
#3      SidekickVersionChecker.getCurrentMinimumPackageVersion (package:sidekick_core/src/sidekick_version_checker.dart:83)
#4      SidekickCommandRunner._checkCliVersionIntegrity (package:sidekick_core/sidekick_core.dart:169)
#5      SidekickCommandRunner.run (package:sidekick_core/sidekick_core.dart:131)
```

File: sidekick_core/lib/src/sidekick_version_checker.dart [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    final lines = pubspec.readAsLinesSync();
+
+    final newVersionConstraint = pinVersion
+        ? newMinimumVersion.canonicalizedVersion
+        : newMinimumVersion.major > 0
+            ? '^${newMinimumVersion.canonicalizedVersion}'
+            : "'>=${newMinimumVersion.canonicalizedVersion} <1.0.0'";
+
+    final index = lines.indexWhere((it) => it.startsWith('  $package:'));
+    lines[index] = '  $package: $newVersionConstraint';
```
Comment (passsy) [inline sidekick_core/lib/src/sidekick_version_checker.dart:55]:
```
Got dependencies!
Sidekick Warning: No Dart SDK set. Please set it in `initializeSidekick(dartSdkPath: 'path/to/sdk')`
Updating sidekick CLI aloh from version 0.0.0 to 0.11.1 ...
There was an error updating sidekick CLI aloh from version 0.0.0 to 0.11.1.
Unhandled exception:
RangeError (index): Invalid value: Not in inclusive range 0..23: -1
#0      List._setIndexed (dart:core-patch/growable_array.dart:273:49)
#1      List.[]= (dart:core-patch/growable_array.dart:268:5)
#2      SidekickVersionChecker.updateVersionConstraint (package:sidekick_core/src/sidekick_version_checker.dart:55:10)
#3      main (package:sidekick_core/src/update_sidekick_cli.dart:53:28)
#4      main (file:///Users/pascalwelsch/Projects/phntmxyz/Alohomora/aloh_sidekick/build/update.dart:3:16)
#5      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:32)
#6      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
Sidekick Warning: No Dart SDK set. Please set it in `initializeSidekick(dartSdkPath: 'path/to/sdk')`
```

pubspec:

```
dependencies:
  csv: ^5.0.0
  dcli: any
  intl: ^0.17.0
  sidekick_core: '>=0.11.1 <1.0.0'
  timeago: ^3.0.2
```

Comment (Giuspepe) [reply sidekick_core/lib/src/sidekick_version_checker.dart]: This should be fixed now, I had a lot of fun writing the regex for my solution :D I've added some tests for these edge cases as well

File: sidekick_core/lib/sidekick_core.dart [hunk 8]
```diff
@@ -126,12 +127,53 @@ class SidekickCommandRunner<T> extends CommandRunner<T> {
 
     final unmount = mount();
     try {
-      final result = await super.run(args);
+      // throw if the user didn't fully update their CLI
+      _checkCliVersionIntegrity();
```
[the remark this answers is not in the document]
Comment (Giuspepe) [reply sidekick_core/lib/sidekick_core.dart]: only a warning is printed now

File: sidekick/lib/src/init/init_command.dart [hunk 9]
```diff
@@ -186,6 +186,11 @@ class InitCommand extends Command {
           .any((package) => package.isFlutterPackage),
       entrypointLocation: entrypoint,
       packageLocation: cliPackage,
+      sidekickCliVersion: version,
+      // We can't read the pubspec.yaml to get the version because the global
+      // sidekick executable is compiled and thus does not have access to pubspec.yaml because it isn't bundled with the compiled executable
+      // Instead we could use an automatically generated source file which contains the version information
+      // See e.g. https://stackoverflow.com/a/69048343/9905602 or how mason_cli implements this in a version.dart file https://github.com/felangel/mason/pull/538/files
```
Comment (passsy) [inline sidekick/lib/src/init/init_command.dart:193]: Comment is now obsolete

File: sidekick_core/pubspec.yaml [hunk 10]
```diff
@@ -20,6 +21,8 @@ dependencies:
 
 dev_dependencies:
   lint: ^1.5.0
+  # transitive from dcli
+  pubspec2: ^2.4.1
   sidekick_test:
     path: ../sidekick_test
```
Comment (passsy) [inline sidekick_core/pubspec.yaml:27]: Unrelated question: Are path `dev_dependencies` allowed when publishing? I assumed all path dependencies are forbidden

Comment (passsy) [reply sidekick_core/pubspec.yaml:27]: I guess, yes. `dart pub publish --dry-run` doesn't complain

File: sidekick_core/lib/sidekick_core.dart [hunk 11]
```diff
@@ -134,15 +135,77 @@ class SidekickCommandRunner<T> extends CommandRunner<T> {
...
+
+    String? command;
     try {
-      final result = await super.run(args);
+      final parsedArgs = parse(args);
+      command = parsedArgs.command?.name;
+      final result = await super.runCommand(parsedArgs);
       return result;
     } finally {
+      if (_isUpdateCheckEnabled && command != 'update') {
```
Comment (passsy) [inline sidekick_core/lib/sidekick_core.dart:146]: what if another command has a subcommand named `update`. Will this run or not?

Comment (Giuspepe) [reply sidekick_core/lib/sidekick_core.dart]: good catch, the check is now more accurate and I added some tests to verify this

File: sidekick_core/lib/sidekick_core.dart [hunk 12]
```diff
@@ -134,15 +135,80 @@ class SidekickCommandRunner<T> extends CommandRunner<T> {
...
     final unmount = mount();
+
+    late final bool isSidekickUpdateCommand;
     try {
-      final result = await super.run(args);
+      final parsedArgs = parse(args);
+      final command = parsedArgs.command;
+      isSidekickUpdateCommand = command?.name == 'sidekick' &&
+          command is ArgResults &&
+          command.command?.name == 'update';
```
Comment (passsy) [inline sidekick_core/lib/sidekick_core.dart:145]: Could we check for the `command` type? That would make it even more resilient and allows users to place the `SidekickCommand` anywhere in the command tree

File: sidekick_core/test/sidekick_command_runner_test.dart [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    addTearDown(() => env['SIDEKICK_ENABLE_UPDATE_CHECK'] = null);
+  });
+
+  final expectedCliVersionIntegrityWarning =
+      'The sidekick_core version is incompatible with the bash scripts '
+      'in /tool and entrypoint because you probably updated the '
+      'sidekick_core dependency of your CLI package manually.\n'
+      'Please run ${cyan('dash sidekick update')} to repair your CLI.';
+  final expectedOutdatedWarning = '${yellow('Update available!')}\n'
+      'Run ${cyan('dash sidekick update')} to update your CLI.';
```
Comment (passsy) [inline sidekick_core/test/sidekick_command_runner_test.dart:22]: I prefer `contains` checks for certain phrases. Those assertions are less likely to break over time

File: sidekick_core/test/sidekick_command_runner_test.dart [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+      },
+      sidekickCoreVersion: "0.0.2",
+      sidekickCliVersion: "0.0.1",
+    );
+  });
+
+  test('prints no warnings when offline and versions match', () async {
+    addTearDown(() => HttpOverrides.global = null);
+    // mock being offline by overriding with a HttpClient which only throws
+    HttpOverrides.global = _ThrowingHttpClient();
```
Comment (passsy) [inline sidekick_core/test/sidekick_command_runner_test.dart:54]: Fine for now. I think in the future we should inject `getLatestDependencyVersion` somehow, so it can be replaced in tests

File: sidekick_core/test/sidekick_command_runner_test.dart [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+    await insideFakeProjectWithSidekick(
+      (tempDir) async {
+        await overrideIoStreams(
+          stderr: () => fakeStderr,
+          body: () async {
+            final runner = initializeSidekick(
+              name: 'dash',
+            );
+
+            await runner.run(['-h']);
```
Comment (passsy) [inline sidekick_core/test/sidekick_command_runner_test.dart:35]: We're doing a lookup at pub.dev, right? That part should be mocked

File: sidekick_core/test/update_command_test.dart [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+            'tool/install.sh',
+            'tool/run.sh',
+            'tool/sidekick_config.sh',
+          ].map(sidekickDir.file);
+
+          for (final file in expectedFilesToGenerate) {
+            expect(
+              !file.existsSync() || file.readAsStringSync().isEmpty,
+              isTrue,
+              reason: '${file.path} exists or is not empty',
```
Comment (passsy) [inline sidekick_core/test/update_command_test.dart:31]: I'm glad you added a reason 👍

File: sidekick_core/lib/src/commands/update_command.dart [hunk 17]
```diff
@@ -0,0 +1,69 @@
+import 'package:sidekick_core/sidekick_core.dart';
+import 'package:sidekick_core/src/version_checker.dart';
+
+/// Updates the sidekick cli
+class UpdateCommand extends Command {
```
Comment (passsy) [inline sidekick_core/lib/src/commands/update_command.dart:5]: Should we add an argument so that the users can specify which version they want to update? Might be handy when the latest version might cause a bug.

File: sidekick_core/lib/src/update_sidekick_cli.dart [hunk 18]
```diff
@@ -0,0 +1,72 @@
+import 'package:sidekick_core/sidekick_core.dart';
+import 'package:sidekick_core/src/version_checker.dart';
+
+Future<void> main(List<String> args) async {
```
Comment (passsy) [inline sidekick_core/lib/src/update_sidekick_cli.dart:4]: add dartdoc

File: sidekick_core/lib/src/commands/update_command.dart [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+  Future<void> run() async {
+    final versionChecker = VersionChecker(Repository.requiredSidekickPackage);
+    // to remember which sidekick_core version the sidekick CLI was generated
+    // with, that sidekick_core version is written into the CLI's pubspec.yaml
+    // at the path ['sidekick', 'cli_version']
+
+    final latestSidekickCoreVersion =
+        await versionChecker.getLatestDependencyVersion('sidekick_core');
+    final currentSidekickCliVersion =
+        versionChecker.getMinimumVersionConstraint(['sidekick', 'cli_version']);
```
Comment (passsy) [inline sidekick_core/lib/src/commands/update_command.dart:22]: Crashing for CLI's that do not contain `sidekic.cli_version` in `pubspec.yaml`. Meaning for all right now

File: sidekick_core/lib/src/commands/update_command.dart (hunk 17 above)
Comment (Giuspepe) [reply sidekick_core/lib/src/commands/update_command.dart:5]: Sounds good 👍

Which version of the `update_sidekick_cli.dart` do we want to download in that case?

1. Always download the version specified by the user
2. Always download the latest available version

E.g.:
- CLI is at `1.0.0` and user specifies to update to `1.1.0` while the latest available version is `1.2.0` (non-breaking).   
- CLI is at `1.0.0` and user specifies to update to `1.1.0` while the latest available version is `2.0.0` (breaking).


I've implemented (1), but I think (2) would be better because the script receives the target version anyways and the advantage of (2) is if the update script had a bug in a certain version, it would be impossible to update to that version.  
A disadvantage of (2) is that it would break when the latest version of the update script has a breaking protocol change (i.e. expects different arguments).  

So maybe it would be best to download the latest available version compatible with the specified version instead.
E.g. 
- CLI is at `1.0.0`, user specifies to update to `1.1.0`, available versions are `[1.0.0, 1.1.0, 1.2.0, 2.0.0]` -> download update_script @ `1.2.0`
- same as above, except CLI is at `0.9.0` -> download update_script @ `1.2.0`

What do you think?

Comment (passsy) [reply sidekick_core/lib/src/commands/update_command.dart:5]: Just for my understanding: 
Right now we update `sidekick_core` to the version provided as argument (vA). Then do `pub get`
Then write and execute the update script of `sidekick_core:A`.

If we have a bug in the upgrade script, a user can always try to update to a different version.
If we would always get the latest update script, a user would be unable to install any version (even old ones that worked in the past) because our new version is buggy or incompatible.

I prefer scenario 1, as implemented

File: sidekick_core/lib/src/commands/update_command.dart (hunk 19 above)
Comment (Giuspepe) [reply sidekick_core/lib/src/commands/update_command.dart]: Fixed it and added a test for it

Comment (passsy) [reply sidekick_core/lib/src/commands/update_command.dart]: Works :)

File: sidekick_core/test/sidekick_command_runner_test.dart (hunk 14 above)
Comment (Giuspepe) [reply sidekick_core/test/sidekick_command_runner_test.dart:54]: I made `VersionChecker` injectable for `SidekickCommandRunner`

File: sidekick_core/test/sidekick_command_runner_test.dart (hunk 15 above)
Comment (Giuspepe) [reply sidekick_core/test/sidekick_command_runner_test.dart:41]: I made `VersionChecker` injectable for `SidekickCommandRunner` so we can fake being offline
